### ISSUE TITLE
UPNP XML parsing crash.

### ISSUE THREAD

**Issue by pchote** [OpenRA/OpenRA#8794]
Reported at https://www.reddit.com/r/openra/comments/3emk67/new_error_openragameexe_has_stopped_working/

```
Red Alert Mod at Version release-20150614
Operating System: Windows (Microsoft Windows NT 6.2.9200.0)
Runtime Version: .NET CLR 4.0.30319.34209
Exception of type System.Xml.XmlException: An error occurred while parsing EntityName. Line 1, position 459.
at System.Xml.XmlTextReaderImpl.Throw(Exception e)
at System.Xml.XmlTextReaderImpl.Throw(String res, String arg)
at System.Xml.XmlTextReaderImpl.ParseEntityName()
at System.Xml.XmlTextReaderImpl.ParseEntityReference()
at System.Xml.XmlTextReaderImpl.Read()
at System.Xml.XmlLoader.LoadNode(Boolean skipOverWhitespace)
at System.Xml.XmlLoader.LoadDocSequence(XmlDocument parentDoc)
at System.Xml.XmlLoader.Load(XmlDocument doc, XmlReader reader, Boolean preserveWhitespace)
at System.Xml.XmlDocument.Load(XmlReader reader)
at System.Xml.XmlDocument.LoadXml(String xml)
at Mono.Nat.Upnp.MessageBase.Decode(UpnpNatDevice device, String message)
at Mono.Nat.Upnp.UpnpNatDevice.DecodeMessageFromResponse(Stream s, Int64 length)
at Mono.Nat.Upnp.UpnpNatDevice.EndMessageInternal(IAsyncResult result)
at Mono.Nat.Upnp.UpnpNatDevice.EndGetAllMappingsInternal(IAsyncResult result)
at System.Net.LazyAsyncResult.Complete(IntPtr userToken)
at System.Net.ContextAwareResult.CaptureOrComplete(ExecutionContext& cachedContext, Boolean returnContext)
at System.Net.ContextAwareResult.FinishPostingAsyncOp()
at System.Net.HttpWebRequest.BeginGetResponse(AsyncCallback callback, Object state)
at Mono.Nat.Upnp.UpnpNatDevice.EndGetAllMappingsInternal(IAsyncResult result)
at System.Net.LazyAsyncResult.Complete(IntPtr userToken)
at System.Net.ContextAwareResult.CaptureOrComplete(ExecutionContext& cachedContext, Boolean returnContext)
at System.Net.ContextAwareResult.FinishPostingAsyncOp()
at System.Net.HttpWebRequest.BeginGetResponse(AsyncCallback callback, Object state)
at Mono.Nat.Upnp.UpnpNatDevice.EndGetAllMappingsInternal(IAsyncResult result)
at System.Net.LazyAsyncResult.Complete(IntPtr userToken)
at System.Net.ContextAwareResult.CaptureOrComplete(ExecutionContext& cachedContext, Boolean returnContext)
at System.Net.ContextAwareResult.FinishPostingAsyncOp()
at System.Net.HttpWebRequest.BeginGetResponse(AsyncCallback callback, Object state)
at Mono.Nat.Upnp.UpnpNatDevice.EndGetAllMappingsInternal(IAsyncResult result)
at System.Net.LazyAsyncResult.Complete(IntPtr userToken)
at System.Net.ContextAwareResult.CaptureOrComplete(ExecutionContext& cachedContext, Boolean returnContext)
at System.Net.ContextAwareResult.FinishPostingAsyncOp()
at System.Net.HttpWebRequest.BeginGetResponse(AsyncCallback callback, Object state)
at Mono.Nat.Upnp.UpnpNatDevice.EndGetAllMappingsInternal(IAsyncResult result)
at System.Net.LazyAsyncResult.Complete(IntPtr userToken)
at System.Net.ContextAwareResult.CaptureOrComplete(ExecutionContext& cachedContext, Boolean returnContext)
at System.Net.ContextAwareResult.FinishPostingAsyncOp()
at System.Net.HttpWebRequest.BeginGetResponse(AsyncCallback callback, Object state)
at Mono.Nat.Upnp.UpnpNatDevice.EndGetAllMappingsInternal(IAsyncResult result)
at System.Net.LazyAsyncResult.Complete(IntPtr userToken)
at System.Net.ContextAwareResult.CaptureOrComplete(ExecutionContext& cachedContext, Boolean returnContext)
at System.Net.ContextAwareResult.FinishPostingAsyncOp()
at System.Net.HttpWebRequest.BeginGetResponse(AsyncCallback callback, Object state)
at Mono.Nat.Upnp.UpnpNatDevice.EndGetAllMappingsInternal(IAsyncResult result)
at System.Net.LazyAsyncResult.Complete(IntPtr userToken)
at System.Net.ContextAwareResult.CaptureOrComplete(ExecutionContext& cachedContext, Boolean returnContext)
at System.Net.ContextAwareResult.FinishPostingAsyncOp()
at System.Net.HttpWebRequest.BeginGetResponse(AsyncCallback callback, Object state)
at Mono.Nat.Upnp.UpnpNatDevice.EndGetAllMappingsInternal(IAsyncResult result)
at System.Net.LazyAsyncResult.Complete(IntPtr userToken)
at System.Net.ContextAwareResult.CaptureOrComplete(ExecutionContext& cachedContext, Boolean returnContext)
at System.Net.ContextAwareResult.FinishPostingAsyncOp()
at System.Net.HttpWebRequest.BeginGetResponse(AsyncCallback callback, Object state)
at Mono.Nat.Upnp.UpnpNatDevice.EndGetAllMappingsInternal(IAsyncResult result)
at System.Net.LazyAsyncResult.Complete(IntPtr userToken)
at System.Net.ContextAwareResult.CaptureOrComplete(ExecutionContext& cachedContext, Boolean returnContext)
at System.Net.ContextAwareResult.FinishPostingAsyncOp()
at System.Net.HttpWebRequest.BeginGetResponse(AsyncCallback callback, Object state)
at Mono.Nat.Upnp.UpnpNatDevice.EndGetAllMappingsInternal(IAsyncResult result)
at System.Net.LazyAsyncResult.Complete(IntPtr userToken)
at System.Net.ContextAwareResult.CaptureOrComplete(ExecutionContext& cachedContext, Boolean returnContext)
at System.Net.ContextAwareResult.FinishPostingAsyncOp()
at System.Net.HttpWebRequest.BeginGetResponse(AsyncCallback callback, Object state)
at Mono.Nat.Upnp.UpnpNatDevice.EndGetAllMappingsInternal(IAsyncResult result)
at System.Net.LazyAsyncResult.WorkerThreadComplete(Object state)
at System.Threading.QueueUserWorkItemCallback.WaitCallback_Context(Object state)
at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
at System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
at System.Threading.ThreadPoolWorkQueue.Dispatch()
at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()
```

**Comment by RoosterDragon** [comment]
tl;dr: To stop this happening, disable network discovery in the game settings.

----

I dig some digging. If a network device sends back an invalid response about port forwarding related things then Mono.Nat will crash our process and there's not much we can do to prevent it from doing so. More details below.

When we call ```NatDevice.GetAllMappings()``` in [UPnP.cs](https://github.com/OpenRA/OpenRA/blob/8114a537608a6d441390f0b90396f19897edb44e/OpenRA.Game/Network/UPnP.cs#L67), this sends some HTTP requests to devices asking for the current mappings.

As the stack trace above shows, a worker thread is used in order to process these web requests asynchronously.

Unfortunately for us, Mono.Nat expects XML back from this web request, and if it doesn't get XML it doesn't bother handling the exception.

However, since the exception occurs on a worker thread, it won't ever make it back to the try-catch we have wrapping the ```NatDevice.GetAllMappings()``` call (this would require Mono.Nat to catch the exception and rethrow it on the right thread for us). Instead it goes unhandled and takes down our entire process as a result.

This is pretty dumb and there's not an awful lot we can do about this since the bug is in in the Mono.Nat library.

This bug can be hit any time a device on the network returns invalid XML for any request - so basically any time we call a method on ```NatDevice```.

I gather from other issues we don't have a lot of options to resolve this:
- We can't handle the error in our code.
- Mono.Nat is dead, so we can't fix it in that.
- We've debated switching to another library in the past to no success: #7140

Fun.

**Comment by pchote** [comment]
We could fork and fix our own copy of Mono.Nat.

**Comment by LipkeGu** [comment]
Extracting this functions would be better than shipping an entire libary

**Comment by LipkeGu** [comment]
I did this for another project and embedd the functions directly in c++ we could also do that ... and implement that cleary in OpenRA

**Comment by chrisforbes** [comment]
+1 for fork-and-fix.

**Comment by Mailaender** [comment]
We use the https://github.com/nterry/Mono.Nat fork which also provides the Nuget package so upstream collaboration is possible here.

**Comment by GraionDilach** [comment]
This isn't a Windows-only crash though, I had this in Linux too.

**Comment by Mailaender** [comment]
@nterry: Have you asked @Xamarin already to gain control over https://github.com/mono/Mono.Nat so we can collaborate on fixing up Mono.NAT all together? Creating yet another fork doesn't sound too sustainable in my eyes also in regard of re-using system libraries when distributing under Linux. https://github.com/OpenRA/OpenRA/issues/8264

**Comment by nterry** [comment]
@Mailaender Yes, but they said that they wouldn't... Im still working on that.
